packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts

```diff
@@ -0,0 +1,232 @@
+}
+
+type Operation =
+  | { readonly op: 'add'; readonly path: string; readonly value: any }
+  | { readonly op: 'remove'; readonly path: string }
+  | { readonly op: 'replace'; readonly path: string; readonly value: any }
+  | { readonly op: 'copy'; readonly path: string; readonly from: string }
+  | { readonly op: 'move'; readonly path: string; readonly from: string }
+  | { readonly op: 'test'; readonly path: string; readonly value: any }
+  ;
+
+/**
+ * Find the sets of patches to apply in a document
+ *
+ * Adjusts paths to be root-relative, which makes it possible to have paths
+ * point outside the patch scope.
+ */
+function findPatches(data: any, patchSource: any): Patch[] {
+  const ret: Patch[] = [];
+  recurse(data, patchSource, []);
+  return ret;
+
+  function recurse(actualData: any, fragment: any, jsonPath: string[]) {
+    if (!fragment) { return; }
+
+    if ('patch' in fragment) {
+      ret.push({
+        description: fragment.patch.description,
+        operations: fragment.patch.operations.map((op: any) => adjustPaths(op, jsonPath)),
+      });
+    } else if ('patch:each' in fragment) {
+      const p = fragment['patch:each'];
+      if (typeof actualData !== 'object') {
+        throw new Error(`Patch ${p.description}: expecting object in data, found '${actualData}'`);
+      }
+      for (const key in actualData) {
+        ret.push({
+          description: `${key}: ${p.description}`,
+          operations: p.operations.map((op: any) => adjustPaths(op, [...jsonPath, key])),
+        });
+      }
+    } else {
+      for (const key of Object.keys(fragment)) {
+        recurse(actualData[key], fragment[key], [...jsonPath, key]);
+      }
+    }
+  }
+
+  function adjustPaths(op: any, jsonPath: string[]): Operation {
+    return {
+      ...op,
+      ...op.path ? { path: adjustPath(op.path, jsonPath) } : undefined,
+      ...op.from ? { from: adjustPath(op.from, jsonPath) } : undefined,
+    };
+  }
+
+  /**
+   * Adjust path
+   *
+   * '$/' means from the root, otherwise interpret as relative path.
+   */
+  function adjustPath(originalPath: string, jsonPath: string[]): string {
+    if (typeof originalPath !== 'string') {
+      throw new Error(`adjustPath: expected string, got ${JSON.stringify(originalPath)}`);
+    }
+    if (originalPath.startsWith('$/')) {
+      return originalPath.substr(1);
+    }
+    return jsonPath.map(p => `/${p}`).join('') + originalPath;
+  }
+}
+
+
+/**
+ * Run this file as a CLI tool, to apply a patch stack from the command line
+ *
+ * Usage: patch-stack <TARGET_FILE> <DIR> [DIR [...]]
+ */
+async function main(args: string[]) {
+  const quiet = eatArg('-q', args) || eatArg('--quiet', args);
+
+  if (args.length < 1) {
+    throw new Error('Usage: patch-stack <DIR> [<FILE>]');
+  }
+
+  const [dir, targetFile] = args;
+
+  if (targetFile) {
+    await resolveStack(targetFile, dir, { quiet });
+  } else {
+    printSorted(await loadStack(dir, { quiet }));
+  }
```

Comment:
Just from a logical flow perspective, I'd probably rewrite this something like:

```suggestion
  const stack = await loadStack(dir, { quiet });
  if (targetFile) {
    writeSorted(targetFile, stack);
  } else {
    printSorted(stack);
  }
```

.. and just get rid of `resolveStack` altogether.